### PR TITLE
Add support to save non-mutable int in StateSnapshot

### PR DESCRIPTION
**Pull request by jingwei99**
Tested locally, saving and restoring rememberSaveable of `non-mutableState` works as expected

### CODE REVIEW

**Comment by jingwei99** [comment]
The focus of this draft is to support save & restore `Int` type (that is `non-mutableState`) across configuration change, that behavior is working as of changes from this PR

After ^, I copied changes from https://github.com/cashapp/redwood/pull/1494 over to test restoring scroll position, but it's not working, @veyndan can you take a look 🙏?

File: redwood-lazylayout-compose/src/commonMain/kotlin/app/cash/redwood/lazylayout/compose/LazyListState.kt [hunk 1]
```diff
@@ -46,4 +47,17 @@ public class LazyListState(
...
+     */
+    public val Saver: Saver<LazyListState, *> = Saver(
+      save = { it.firstVisibleItemIndex },
+      restore = {
+        LazyListState(
+          firstVisibleItemIndex = it,
+        )
+      }
+    )
+  }
```
Comment (jingwei99) [inline redwood-lazylayout-compose/src/commonMain/kotlin/app/cash/redwood/lazylayout/compose/LazyListState.kt]: this is copy-paste from https://github.com/cashapp/redwood/pull/1494, and will be reverted once this PR goes from draft -> ready for review

**Comment by veyndan** [comment]
What doesn't work with the PR? Is it that saving a plain `Int` doesn't work (i.e., ignoring scroll state restoration)? Or is it that this PR does what its supposed to, but when integrating with #1494, scroll state restoration doesn't work?

**Comment by jingwei99** [comment]
> when integrating with #1494, scroll state restoration doesn't work

this^

**Comment by veyndan** [comment]
Cool, so if saving a plain `Int` works, shall we mark this PR as a non-draft and get it merged? Your PR isn't strictly related to scroll state restoration, as it is useful in its own right. We can tackle any scroll state restoration bugs in #1494.

**Comment by jingwei99** [comment]
> Cool, so if saving a plain `Int` works, shall we mark this PR as a non-draft and get it merged? Your PR isn't strictly related to scroll state restoration, as it is useful in its own right. We can tackle any scroll state restoration bugs in #1494.

yep, sounds good to me, will clean up this PR and put it up for review

File: redwood-treehouse/src/commonTest/kotlin/app/cash/redwood/treehouse/StateSnapshotTest.kt [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+
+import androidx.compose.runtime.MutableState
+import androidx.compose.runtime.mutableStateOf
+import assertk.assertThat
+import assertk.assertions.isEqualTo
+import kotlin.test.Test
+import kotlin.test.assertTrue
+import kotlinx.serialization.json.JsonPrimitive
+
+class StateSnapshotTest {
```
Comment (veyndan) [inline redwood-treehouse/src/commonTest/kotlin/app/cash/redwood/treehouse/StateSnapshotTest.kt:27]: Tests! Love it 😄

File: redwood-treehouse/src/commonTest/kotlin/app/cash/redwood/treehouse/StateSnapshotTest.kt [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+    assertThat(valuesMap.entries.size).isEqualTo(4)
+    assertTrue(valuesMap["key1"]!![0] is MutableState<*>)
+    assertThat((valuesMap["key1"]!![0] as MutableState<*>).value).isEqualTo(JsonPrimitive(1))
+
+    assertThat(valuesMap["key2"]).isEqualTo(listOf(1.0))
+
+    assertThat(valuesMap["key3"]!![0] is MutableState<*>)
+    assertThat((valuesMap["key3"]!![0] as MutableState<*>).value).isEqualTo(JsonPrimitive("str"))
+
+    assertThat(valuesMap["key4"]).isEqualTo(listOf("str"))
```
Comment (veyndan) [inline redwood-treehouse/src/commonTest/kotlin/app/cash/redwood/treehouse/StateSnapshotTest.kt:42]: Could these test cases be simplified with the [`containsOnly` function](https://willowtreeapps.github.io/assertk/assertk/assertk.assertions/contains-only.html)?

```kotlin
    assertThat(valuesMap).containsOnly(
        "key1" to listOf(mutableStateOf(JsonPrimitive(1)),
        …
    )
```

File: redwood-treehouse/src/commonTest/kotlin/app/cash/redwood/treehouse/StateSnapshotTest.kt [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+  }
+
+  @Test
+  fun toStateSnapshotWorksAsExpected() {
+    val storedStateSnapshot = storedStateSnapshot()
+    val stateSnapshot = storedStateSnapshot.toStateSnapshot()
+    assertThat(stateSnapshot.content["key1"]!![0]).isEqualTo(Saveable(true, JsonPrimitive(1)))
+    assertThat(stateSnapshot.content["key2"]!![0]).isEqualTo(Saveable(false, JsonPrimitive(1)))
+    assertThat(stateSnapshot.content["key3"]!![0]).isEqualTo(Saveable(true, JsonPrimitive("str")))
+    assertThat(stateSnapshot.content["key4"]!![0]).isEqualTo(Saveable(false, JsonPrimitive("str")))
```
Comment (veyndan) [inline redwood-treehouse/src/commonTest/kotlin/app/cash/redwood/treehouse/StateSnapshotTest.kt]: Same idea with these test cases

File: redwood-treehouse/src/commonMain/kotlin/app/cash/redwood/treehouse/StateSnapshot.kt [hunk 5]
```diff
@@ -18,20 +18,28 @@ package app.cash.redwood.treehouse
...
-  public fun toValuesMap(): Map<String, List<Any?>>? {
+  public fun toValuesMap(): Map<String, List<Any?>> {
     return content.mapValues { entry ->
-      entry.value.map { mutableStateOf(it.fromJsonElement()) }
+      entry.value.map {
+        if (it.isMutableState) {
+          mutableStateOf(it.value)
+        } else {
+          it.value.fromJsonElement()
+        }
```
Comment (veyndan) [inline redwood-treehouse/src/commonMain/kotlin/app/cash/redwood/treehouse/StateSnapshot.kt:41]: We could remove the need for the `Saveable` wrapper class if we wrote an adapter to convert between `MutableState` and `JsonElement`. Perhaps it isn't worth the effort though, considering we'll change up this mechanism in #1384.

File: redwood-treehouse/src/commonMain/kotlin/app/cash/redwood/treehouse/StateSnapshot.kt [hunk 6]
```diff
@@ -68,11 +77,9 @@ private fun Any?.toJsonElement(): JsonElement {
 private fun JsonElement?.fromJsonElement(): Any {
   return when (this) {
     is JsonPrimitive -> {
-      if (this.isString) {
-        return content
-      }
-      return booleanOrNull ?: doubleOrNull ?: error("unexpected type: $this")
-      // TODO add other primitive types (double, float, long) when needed
+      if (this.isString) { return content }
```
Comment (veyndan) [inline redwood-treehouse/src/commonMain/kotlin/app/cash/redwood/treehouse/StateSnapshot.kt]: This looks a bit odd having the braces on a single line. Can we revert it, or do `if (this.isString) return content`?

**Review by veyndan — APPROVED** [review]

File: redwood-treehouse/src/commonTest/kotlin/app/cash/redwood/treehouse/StateSnapshotTest.kt (hunk 3 above)
Comment (jingwei99) [reply redwood-treehouse/src/commonTest/kotlin/app/cash/redwood/treehouse/StateSnapshotTest.kt:42]: `containsOnly` errors out: 

`expected to contain only:<{"key1"=[MutableState(value=1)@916835004], "key2"=[1.0], "key3"=[MutableState(value="str")@2108297149], "key4"=["str"]}> but was:<{"key1"=[MutableState(value=1)@1112737073], "key2"=[1.0], "key3"=[MutableState(value="str")@1513867245], "key4"=["str"]}>
 elements not found:<{"key1"=[MutableState(value=1)@916835004], "key3"=[MutableState(value="str")@2108297149]}>
 extra elements found:<{"key1"=[MutableState(value=1)@1112737073], "key3"=[MutableState(value="str")@1513867245]}>`

for the `MutableState` ones, `containsOnly` doesn't unwrap and compare the values (it compares the instances)

File: redwood-treehouse/src/commonTest/kotlin/app/cash/redwood/treehouse/StateSnapshotTest.kt (hunk 4 above)
Comment (jingwei99) [reply redwood-treehouse/src/commonTest/kotlin/app/cash/redwood/treehouse/StateSnapshotTest.kt]: this one works out

**Pull request merged**